### ISSUE TITLE
Dependency injection fails to resolve token from argument type if it is a union type with undefined

### ISSUE THREAD

**Issue by Yogu** [angular/angular#25544]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior

If constructor arguments include `null` or `undefined` union options, Angular no longer detects the injection token from the argument type and throws the error `Can't resolve all parameters for AppComponent: (?).`.

## Expected behavior

It would be good if the union options `undefined` and `null` would be ignored. (Or even would trigger / allow `@Optional`?)

## Minimal reproduction of the problem with instructions

Write a constructor like this (with `strictNullChecksEnabled`):

```typescript
export class AppComponent  {
  constructor(@Optional() http: HttpClient|123) {

  }
}
```

## What is the motivation / use case for changing the behavior?

With `strictNullChecks` enabled, it would be good to mark optional injection dependencies as `| undefined` to make sure the implementation can cope with the dependency being missing.

However, a workaround is to just provide the `@Inject()` decorator.

## Environment

<pre><code>
Angular version: 6.1.0

Browser:
- [x] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.11.3  
- Platform:  Windows

Others:

</code></pre>

**Comment by alxhub** [comment]
Likely we will fix this in Ivy.

**Comment by trotyl** [comment]
@alxhub This is the expected behavior in TypeScript: https://github.com/Microsoft/TypeScript/issues/18509, it won't be fixed until JIT support being dropped.

**Comment by CSchulz** [comment]
Will this also cover the case with correct type declarations?

> The DI framework provides null if the dependency is not found.

So I am expecting the value can be null or in other words will be there an enforcement for the developer to use an "nullable" value container?


```
export class AppComponent  {
  constructor(@Optional() http: HttpClient | null) {

  }
}
```

**Comment by alxhub** [comment]
This was fixed in #27552
